src/operator/tensor/histogram.cc

```diff
@@ -0,0 +1,163 @@
+
+  Tensor<cpu, 1, int> bin_indices =
+    ctx.requested[0].get_space_typed<cpu, 1, int>(Shape1(in_data.Size()), s);
+  const int bin_cnt = out_data.Size();
+  MSHADOW_TYPE_SWITCH(in_data.type_flag_, DType, {
+    if (has_cnt) {
+      float max = param.range.value()[1];
+      float min = param.range.value()[0];
+      float width = (max - min) / bin_cnt;
+      Kernel<ComputeBinKernel, cpu>::Launch(
+        s, in_data.Size(), in_data.dptr<DType>(), bin_indices.dptr_,
+        bin_cnt, width, min, max);
+      Kernel<FillBinBoundsKernel, cpu>::Launch(
+        s, bin_cnt+1, out_bins.dptr<DType>(), bin_cnt, min, max);
+    } else {
+      Kernel<ComputeBinKernel, cpu>::Launch(
+        s, in_data.Size(), in_data.dptr<DType>(), bin_indices.dptr_, bin_bounds.dptr<DType>(),
+        bin_cnt);
+      Kernel<op_with_req<mshadow_op::identity, kWriteTo>, cpu>::Launch(
+        s, bin_bounds.Size(), out_bins.dptr<DType>(), bin_bounds.dptr<DType>());
+    }
+  });
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, CType, {
+    Kernel<set_zero, cpu>::Launch(s, bin_cnt, out_data.dptr<CType>());
+    ComputeHistogram(bin_indices.dptr_, out_data.dptr<CType>(), in_data.Size());
+  });
+}
+
+template<>
+void HistogramBackwardImpl<cpu>(const OpContext& ctx,
+                                const nnvm::NodeAttrs& attrs,
+                                const TBlob& out_grad,
+                                const TBlob& in_data,
+                                const TBlob& bin_bounds,
+                                const TBlob& out_data,
+                                const TBlob& in_grad) {
+  LOG(FATAL) << "Histogram Backward not implemented yet";
+}
+
+DMLC_REGISTER_PARAMETER(HistogramParam);
+
+NNVM_REGISTER_OP(_histogram)
+.describe(R"code(This operators implements the histogram function.
+
+Example::
+  x = [[0, 1], [2, 2], [3, 4]]
+  histo, bin_edges = histogram(data=x, bin_bounds=[], bin_cnt=5, range=(0,5))
+  histo = [1, 1, 2, 1, 1]
+  bin_edges = [0., 1., 2., 3., 4.]
+  histo, bin_edges = histogram(data=x, bin_bounds=[0., 2.1, 3.])
+  histo = [4, 1]
+
+)code" ADD_FILELINE)
+.set_attr_parser(ParamParser<HistogramParam>)
+.set_num_inputs(2)
+.set_num_outputs(2)
+.set_attr<nnvm::FListInputNames>("FListInputNames",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::string>{"data", "bins"};
+  })
+.set_attr<nnvm::FInferShape>("FInferShape", HistogramOpShape)
+.set_attr<nnvm::FInferType>("FInferType", HistogramOpType)
+.set_attr<FCompute>("FCompute<cpu>", HistogramOpForward<cpu>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{"_backward_histogram"})
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::pair<int, int> >{};
+  })
+.add_argument("data", "NDArray-or-Symbol", "Input ndarray")
+.add_argument("bins", "NDArray-or-Symbol", "Input ndarray")
+.add_arguments(HistogramParam::__FIELDS__());
+
+NNVM_REGISTER_OP(_backward_histogram)
```

Comment:
Historgram op is not differentiable. Don't register backward op.

Reply:
Sure, will get rid of it.